File: today/parser.py
from typing import Sequence, Tuple, List, Optional
from datetime import date
import re
from more_itertools import windowed

from today.task import (
    AssignmentAttribute,
    PriorityAttribute,
    Task,
    Heading,
    TaskAttributes,
    TaskTitle,
)

task_attr_re = re.compile(r"\[(?P<prefix>(.:|@|!))(?P<value>.*?)\]\s?")
task_re = re.compile(r"^- \[[ xX]\] ")
subtask_re = re.compile(r"^[ \t]+- \[[ xX]\] ")


def parse_heading(s: str) -> Heading:
    for i in range(len(s)):
        if s[i] == " ":
            return Heading(level=i, name=s[i + 1 :])
        elif s[i] == "#":
            continue
        else:
            raise ValueError(f"Malformed heading {s}")
    raise ValueError("This should never happen")


# Given the current state of headings in [headings_stack] and a new heading [heading_raw],
# validate that the new heading has the correct indentation and return a new heading stack
def handle_headings_stack(headings_stack: List[str], heading_raw: str) -> List[str]:
    heading = parse_heading(heading_raw)
    last_level = len(headings_stack)
    if heading.level > last_level:  # deeper into hierarchy
        if heading.level is not last_level + 1:
            raise ValueError(f"Heading {heading} nested too deep")
    elif heading.level == last_level:  # same hierarchy level, different heading
        headings_stack.pop()
    else:  # higher hierarchy level, pop off current level and upper level
        for _ in range(last_level - heading.level + 1):
            headings_stack.pop()
    headings_stack.append(heading.name)
    return headings_stack


def md_checkbox(s: str) -> Optional[bool]:
    # None = not a checkbox, True = checked, False = unchecked
    if s.startswith("[ ]"):
        return False
    elif s.startswith("[x]") or s.startswith("[X]"):
        return True
    else:
        return None


# Mutates the fields of [task_attr] based on a raw attribute string (prefix + value)
# of the form [d:<date>] (prefix='d:', value='<date>') or [@person] or [!2]
# If the prefix or value are malformed, return an error message
def assign_task_attr(
    prefix: str, value: str, task_attr: TaskAttributes, today: date
) -> None | str:
    if prefix == "@":
        # This is an assignment attribute
        task_attr.assn_attr = AssignmentAttribute(value)
        return
    elif prefix == "!":
        # This is a priority attribute
        task_attr.priority_attr = PriorityAttribute(int(value))
        return
    else:
        # This must be a date attribute
        prefix = prefix[0]  # the raw prefix passed is of the form 'd:'
        date_raw = value
        date_value: date
        if date_raw == "t":
            date_value = today
        else:
            date_split = [int(d) for d in date_raw.split("/")]
            if len(date_split) == 3:  # month / day / year
                date_value = date(
                    year=date_split[2], month=date_split[0], day=date_split[1]
                )
            elif len(date_split) == 2:  # month / day (year is implicitly today's year)
                date_value = date(
                    year=today.year, month=date_split[0], day=date_split[1]
                )
            else:
                return f"Date attribute value '{value}' is improperly formatted"
        if prefix == "c":
            task_attr.date_attr.created_date = date_value
        elif prefix == "d":
            task_attr.date_attr.due_date = date_value
        elif prefix == "r":
            task_attr.date_attr.reminder_date = date_value
        elif prefix == "f":
            task_attr.date_attr.finished_date = date_value
        else:
            return f"Date attribute prefix '{prefix}' isn't recognized"
        return


def extract_task_attrs(
    raw_task_title: str, today: date
) -> Tuple[TaskAttributes, TaskTitle]:
    task_attr = TaskAttributes()

    # Find all matches for the attribute regex and parse each one while mutating [task_attr]
    task_attr_matches = list(re.finditer(task_attr_re, raw_task_title))
    if len(task_attr_matches) == 0:
        # short circuit when there are no attributes to parse
        return task_attr, raw_task_title
    for match in task_attr_matches:
        prefix = match.group("prefix")
        value = match.group("value")
        error_msg = assign_task_attr(prefix, value, task_attr, today)
        if error_msg is not None:
            raise RuntimeError(
                f"An error was encountered when parsing the task title '{raw_task_title}'. Error: {error_msg}"
            )

    # Strip all the task attribute matches from the [raw_task_title]
    # Do this efficiently by first computing the substring indices we need, then constructing the new string
    match_spans: List[Tuple[int, int]] = [m.span(0) for m in task_attr_matches]
    not_match_spans = (
        [(0, match_spans[0][0])]
        + [
            (span1[1], span2[0])
            for (span1, span2) in windowed(match_spans, 2)
            if span1 is not None and span2 is not None
        ]
        + [(match_spans[-1][1], len(raw_task_title))]
    )
    task_title = ""
    for span in not_match_spans:
        task_title = task_title + raw_task_title[span[0] : span[1]]
    return task_attr, task_title.rstrip()


def parse_task_title(title: str, today: date) -> Task:
    task_attr, task_title = extract_task_attrs(title, today)
    t = Task(title=task_title, attrs=task_attr)
    return t


def parse_markdown(md: Sequence[str], today: date = date.today()) -> List[Task]:
    headings_stack: List[str] = []
    current_task: Optional[Task] = None
    tasks: List[Task] = []
    for i, line in enumerate(md):
        if line.startswith("#"):  # This is a heading
            headings_stack = handle_headings_stack(headings_stack, line)
            # Headings terminate any task already being parsed
            if current_task is not None:
                tasks.append(current_task)
                current_task = None
        elif task_re.match(line):  # This is a Markdown checkbox (a task)
            task_status = md_checkbox(line[len("- ") :])
            if task_status is not None:
                if current_task is not None:
                    tasks.append(current_task)
                current_task = parse_task_title(line[len("- [ ] ") :], today)
                current_task.path = headings_stack.copy()
                current_task.done = task_status
                current_task.line_number = i + 1
            else:  # Malformed Markdown checkbox
                raise ValueError(f"Malformed Markdown checkbox on line {i}: {line}")
        elif (match := subtask_re.match(line)) is not None:
            if current_task is None:
                raise ValueError(
                    f"Encountered subtask without a main task on line {i}: {line}"
                )
            subtask_status = md_checkbox(line[line.index("[") :])
            assert subtask_status is not None  # The checkbox must not be malformed
            subtask = parse_task_title(line[match.end(0) :], today)
            subtask.path = headings_stack.copy()
            subtask.done = subtask_status
            subtask.line_number = i + 1
            subtask.attrs.merge_attributes(current_task.attrs)
            current_task.subtasks.append(subtask)
        elif len(line) == 0 and current_task is None:
            continue
        else:
            if current_task is None:  # Unparsed text right after a header
                continue
                # raise ValueError(f"Encountered description not associated with a task on line {i}: {line}")
            current_task.description = (
                current_task.description + "\n" + line
            )  # This is part of the description of a current task

    if current_task is not None:
        tasks.append(current_task)

    # Post-process descriptions - remove trailing or leading newlines and spaces
    for i in range(len(tasks)):
        tasks[i].description = tasks[i].description.strip("\n ")
    return tasks

File: src/parser.rs
use crate::task::{AssignmentAttribute, Heading, PriorityAttribute, Task, TaskAttributes};
use chrono::{NaiveDate, Datelike};
use regex::Regex;
use std::sync::OnceLock;

/// Get the task attribute regex (compiled once)
fn task_attr_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| Regex::new(r"\[(?P<prefix>(.:|@|!))(?P<value>.*?)\]\s?").unwrap())
}

/// Get the task line regex
fn task_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| Regex::new(r"^- \[[ xX]\] ").unwrap())
}

/// Get the subtask line regex
fn subtask_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| Regex::new(r"^[ \t]+- \[[ xX]\] ").unwrap())
}

/// Parse a Markdown heading line
pub fn parse_heading(s: &str) -> Result<Heading, String> {
    let mut level = 0;
    let chars: Vec<char> = s.chars().collect();
    
    for (i, &ch) in chars.iter().enumerate() {
        if ch == '#' {
            level = i + 1;
        } else if ch == ' ' {
            return Ok(Heading {
                level,
                name: s[i + 1..].to_string(),
            });
        } else {
            return Err(format!("Malformed heading: {}", s));
        }
    }
    
    Err("This should never happen".to_string())
}

/// Handle the heading stack when encountering a new heading
pub fn handle_headings_stack(headings_stack: &mut Vec<String>, heading_raw: &str) -> Result<(), String> {
    let heading = parse_heading(heading_raw)?;
    let last_level = headings_stack.len();
    
    if heading.level > last_level {
        if heading.level != last_level + 1 {
            return Err(format!("Heading {:?} nested too deep", heading));
        }
    } else if heading.level == last_level {
        headings_stack.pop();
    } else {
        for _ in 0..(last_level - heading.level + 1) {
            headings_stack.pop();
        }
    }
    
    headings_stack.push(heading.name);
    Ok(())
}

/// Check if a line is a Markdown checkbox and return its status
pub fn md_checkbox(s: &str) -> Option<bool> {
    if s.starts_with("[ ]") {
        Some(false)
    } else if s.starts_with("[x]") || s.starts_with("[X]") {
        Some(true)
    } else {
        None
    }
}

/// Parse and assign a task attribute
pub fn assign_task_attr(
    prefix: &str,
    value: &str,
    task_attr: &mut TaskAttributes,
    today: NaiveDate,
) -> Result<(), String> {
    if prefix == "@" {
        task_attr.assn_attr = Some(AssignmentAttribute {
            assigned_to: value.to_string(),
        });
        Ok(())
    } else if prefix == "!" {
        let priority = value.parse::<i32>()
            .map_err(|_| format!("Invalid priority value: {}", value))?;
        task_attr.priority_attr = Some(PriorityAttribute { priority });
        Ok(())
    } else {
        // Date attribute
        let prefix_char = prefix.chars().next().unwrap();
        let date_value = if value == "t" {
            today
        } else {
            parse_date(value, today)?
        };
        
        match prefix_char {
            'c' => task_attr.date_attr.created_date = Some(date_value),
            'd' => task_attr.date_attr.due_date = Some(date_value),
            'r' => task_attr.date_attr.reminder_date = Some(date_value),
            'f' => task_attr.date_attr.finished_date = Some(date_value),
            _ => return Err(format!("Date attribute prefix '{}' isn't recognized", prefix_char)),
        }
        
        Ok(())
    }
}

/// Parse a date string (M/D/Y or M/D)
fn parse_date(date_str: &str, today: NaiveDate) -> Result<NaiveDate, String> {
    let parts: Vec<&str> = date_str.split('/').collect();
    
    match parts.len() {
        3 => {
            let month = parts[0].parse::<u32>()
                .map_err(|_| format!("Invalid month: {}", parts[0]))?;
            let day = parts[1].parse::<u32>()
                .map_err(|_| format!("Invalid day: {}", parts[1]))?;
            let year = parts[2].parse::<i32>()
                .map_err(|_| format!("Invalid year: {}", parts[2]))?;
            
            NaiveDate::from_ymd_opt(year, month, day)
                .ok_or_else(|| format!("Invalid date: {}/{}/{}", month, day, year))
        }
        2 => {
            let month = parts[0].parse::<u32>()
                .map_err(|_| format!("Invalid month: {}", parts[0]))?;
            let day = parts[1].parse::<u32>()
                .map_err(|_| format!("Invalid day: {}", parts[1]))?;
            
            NaiveDate::from_ymd_opt(today.year(), month, day)
                .ok_or_else(|| format!("Invalid date: {}/{}", month, day))
        }
        _ => Err(format!("Date attribute value '{}' is improperly formatted", date_str)),
    }
}

/// Extract task attributes from a raw task title
pub fn extract_task_attrs(raw_task_title: &str, today: NaiveDate) -> Result<(TaskAttributes, String), String> {
    let mut task_attr = TaskAttributes::default();
    let regex = task_attr_regex();
    
    let matches: Vec<_> = regex.find_iter(raw_task_title).collect();
    
    if matches.is_empty() {
        return Ok((task_attr, raw_task_title.to_string()));
    }
    
    // Parse each attribute
    for cap in regex.captures_iter(raw_task_title) {
        let prefix = cap.name("prefix").unwrap().as_str();
        let value = cap.name("value").unwrap().as_str();
        
        assign_task_attr(prefix, value, &mut task_attr, today)
            .map_err(|e| format!("Error parsing task title '{}': {}", raw_task_title, e))?;
    }
    
    // Remove all attribute matches from the title
    let mut result = String::new();
    let mut last_end = 0;
    
    for m in matches {
        result.push_str(&raw_task_title[last_end..m.start()]);
        last_end = m.end();
    }
    result.push_str(&raw_task_title[last_end..]);
    
    Ok((task_attr, result.trim_end().to_string()))
}

/// Parse a task title line
pub fn parse_task_title(title: &str, today: NaiveDate) -> Result<Task, String> {
    let (attrs, task_title) = extract_task_attrs(title, today)?;
    Ok(Task {
        title: task_title,
        attrs,
        ..Task::default()
    })
}

/// Parse markdown lines into tasks
pub fn parse_markdown(md: &[String], today: NaiveDate) -> Result<Vec<Task>, String> {
    let mut headings_stack: Vec<String> = Vec::new();
    let mut current_task: Option<Task> = None;
    let mut tasks: Vec<Task> = Vec::new();
    
    for (i, line) in md.iter().enumerate() {
        if line.starts_with('#') {
            // This is a heading
            handle_headings_stack(&mut headings_stack, line)?;
            
            // Headings terminate any task being parsed
            if let Some(task) = current_task.take() {
                tasks.push(task);
            }
        } else if task_regex().is_match(line) {
            // This is a task checkbox
            let checkbox_content = &line[2..]; // Skip "- "
            let task_status = md_checkbox(checkbox_content)
                .ok_or_else(|| format!("Malformed Markdown checkbox on line {}: {}", i + 1, line))?;
            
            if let Some(task) = current_task.take() {
                tasks.push(task);
            }
            
            let task_content = &line[6..]; // Skip "- [ ] "
            let mut task = parse_task_title(task_content, today)?;
            task.path = headings_stack.clone();
            task.done = task_status;
            task.line_number = i + 1;
            
            current_task = Some(task);
        } else if let Some(mat) = subtask_regex().find(line) {
            // This is a subtask
            let current = current_task.as_mut()
                .ok_or_else(|| format!("Encountered subtask without a main task on line {}: {}", i + 1, line))?;
            
            let checkbox_start = line.find('[').unwrap();
            let checkbox_content = &line[checkbox_start..];
            let subtask_status = md_checkbox(checkbox_content)
                .ok_or_else(|| format!("Malformed subtask checkbox on line {}: {}", i + 1, line))?;
            
            let subtask_content = &line[mat.end()..];
            let mut subtask = parse_task_title(subtask_content, today)?;
            subtask.path = headings_stack.clone();
            subtask.done = subtask_status;
            subtask.line_number = i + 1;
            subtask.attrs.merge_attributes(&current.attrs);
            
            current.subtasks.push(subtask);
        } else if line.is_empty() && current_task.is_none() {
            continue;
        } else if let Some(task) = current_task.as_mut() {
            // This is part of the description
            if !task.description.is_empty() {
                task.description.push('\n');
            }
            task.description.push_str(line);
        }
    }
    
    if let Some(task) = current_task {
        tasks.push(task);
    }
    
    // Post-process descriptions
    for task in &mut tasks {
        task.description = task.description.trim().to_string();
    }
    
    Ok(tasks)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_heading() {
        assert_eq!(
            parse_heading("# Title").unwrap(),
            Heading { level: 1, name: "Title".to_string() }
        );
        assert_eq!(
            parse_heading("### **title 2**").unwrap(),
            Heading { level: 3, name: "**title 2**".to_string() }
        );
        assert!(parse_heading("bare line").is_err());
    }

    #[test]
    fn test_md_checkbox() {
        assert_eq!(md_checkbox("[ ]"), Some(false));
        assert_eq!(md_checkbox("[x]"), Some(true));
        assert_eq!(md_checkbox("[X]"), Some(true));
        assert_eq!(md_checkbox("invalid"), None);
    }
}
